### PR TITLE
add isBreakStatement and pruifying a few is-Methods

### PR DESCRIPTION
**Pull request by UplinkCoder**
Hi,

This is needed for the same reason as #5920.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
@WalterBright could you please merge this ?

**Comment by WalterBright** [comment]
The form that should be used is like in dsymbol.d:
```
inout(TemplateInstance) isTemplateInstance() inout
{
    return null;
}
```
because then they can be used in const-correct functions. Adding `pure nothrow` is also a good idea.

**Comment by UplinkCoder** [comment]
I see. Thanks.

**Comment by UplinkCoder** [comment]
@WalterBright I added inout pure and nothrow.

File: src/statement.d [hunk 1]
```diff
@@ -1324,38 +1329,38 @@ extern (C++) class CompoundStatement : Statement
         return statements;
     }
 
-    override final ReturnStatement isReturnStatement()
+    override final inout(ReturnStatement) isReturnStatement() inout nothrow pure
```
Comment (WalterBright) [inline src/statement.d]: Can't just pick one of the statements and call it "is" a return statement. It isn't.  This works completely unlike all the other dynamic casting virtual functions.

That would be "has" a return statement.

Comment (UplinkCoder) [inline src/statement.d]: I can imagine why the code was structured that way ... 
To have an easy determination if a block returns or not.
I do agree that it should not be structured that way.

However to be pragmatic would you accept this PR with a FIXME note ?

Comment (WalterBright) [inline src/statement.d]: I don't see a reason to not fix it now.

Comment (UplinkCoder) [inline src/statement.d]: Fine. But I'll open another PR for that. It is outside of the scope for this one.

Comment (WalterBright) [inline src/statement.d]: How is it outside the scope of this? The function does not follow the convention used everywhere else for dynamic casting.

Comment (UplinkCoder) [inline src/statement.d]: This is a purely additive PR.
Without the goal of changing behavior anywhere.

**Comment by UplinkCoder** [comment]
@WalterBright do you see now why I said it's not in scope ?
Shall I open another PR for the second commit  or squash it?

**Comment by UplinkCoder** [comment]
This fix sure has alot of fallout...

**Comment by UplinkCoder** [comment]
@WalterBright could you merge this as it. I will follow up on the isReturnStatement issue later.

**Comment by yebblies** [comment]
Auto-merge toggled on

**Comment by UplinkCoder** [comment]
Thanks @yebblies.

**Pull request merged**